ldms/src/sampler/zfs_topvdevs/zfs_topvdevs.c

```diff
@@ -0,0 +1,440 @@
+ *  free
+ *  size
+ *  read_bytes
+ *  read_errors
+ *  read_ops
+ *  write_bytes
+ *  write_errors
+ *  write_ops
+ *  checksum_errors
+ *  fragmentation
+ *  init_errors
+ */
+static int get_vdev_stats(nvlist_t *nvroot, const char *pool_name,
+    zpool_handle_t *zhp)
+{
+        uint_t c;
+        vdev_stat_t *vs;
+        char *vdev_name = NULL;
+	ldms_mval_t record_instance;
+        int rc=0; /*return code*/
+
+	vdev_name = zpool_vdev_name(g_zfs, zhp, nvroot, VDEV_NAME_TYPE_ID);
+
+	if (nvlist_lookup_uint64_array(	nvroot,
+					ZPOOL_CONFIG_VDEV_STATS,
+					(uint64_t **)&vs, &c) != 0) {
+                rc = 1;
+        }
+
+        record_instance = ldms_record_alloc(sampler_base->set,
+                                            index_store.vdev_rec_idx);
+
+        if (record_instance == NULL) {
+                log_fn(LDMSD_LDEBUG, SAMP": ldms_record_alloc() failed, resizing metric set\n");
+                resize_metric_set();
+		record_instance = ldms_record_alloc(sampler_base->set,
+				index_store.vdev_rec_idx);
+		if (record_instance == NULL)
+		  rc = 2;
+        }
+
+	if (rc == 0) {
+		rc = ldms_list_append_record(sampler_base->set, list_handle,
+						record_instance);
+
+		/* zpoolname    0 */
+		ldms_record_array_set_str(record_instance, vdev_metric_ids[0],
+				pool_name);
+	        /* zpool state  1 */
+		ldms_record_array_set_str(record_instance, vdev_metric_ids[1],
+				zpool_state_to_name((vdev_state_t)vs->vs_state, (vdev_aux_t)vs->vs_aux));
+	        /* vdevname     2 */
+		ldms_record_array_set_str(record_instance, vdev_metric_ids[2],
+						vdev_name);
+	        /* alloc        3 */
+		ldms_record_set_u64(record_instance, vdev_metric_ids[3], vs->vs_alloc);
+	        /* free         4 */
+		ldms_record_set_u64(record_instance, vdev_metric_ids[4], vs->vs_space - vs->vs_alloc);
+		/* size         5 */
+		ldms_record_set_u64(record_instance, vdev_metric_ids[5], vs->vs_space);
+		/* read_bytes   6 */
+		ldms_record_set_u64(record_instance, vdev_metric_ids[6], vs->vs_bytes[ZIO_TYPE_READ]);
+		/* iread_errors 7 */
+		ldms_record_set_u64(record_instance, vdev_metric_ids[7], vs->vs_read_errors);
+		/* read_ops     8 */
+		ldms_record_set_u64(record_instance, vdev_metric_ids[8], vs->vs_ops[ZIO_TYPE_READ]);
+		/* write_bytes  9 */
+		ldms_record_set_u64(record_instance, vdev_metric_ids[9], vs->vs_bytes[ZIO_TYPE_WRITE]);
+		/* write_errors 10 */
+		ldms_record_set_u64(record_instance, vdev_metric_ids[10], vs->vs_write_errors);
+		/* write_ops    11 */
+		ldms_record_set_u64(record_instance, vdev_metric_ids[11], vs->vs_ops[ZIO_TYPE_WRITE]);
+		/* checksum errors 12 */
+		ldms_record_set_u64(record_instance, vdev_metric_ids[12], vs->vs_checksum_errors);
+		/* fragmentation 13 */
+		ldms_record_set_u64(record_instance, vdev_metric_ids[13], vs->vs_fragmentation);
+		/* initialization errors 14 */
+		ldms_record_set_u64(record_instance, vdev_metric_ids[14], vs->vs_initialize_errors);
+	}
+
+	free(vdev_name);
+	vdev_name = NULL;
```

Comment:
`vdev_name` goes out of scope here anyway so I'd dropping this assignment to NULL.